lib/src/main/scala/spinal/lib/sim/bus/wishbone/WishboneMonitor.scala

```diff
@@ -16,17 +16,28 @@ object WishboneMonitor{
   */
 class WishboneMonitor(bus: Wishbone, clockdomain: ClockDomain){
   val busStatus = WishboneStatus(bus)
-  val callbacks = ArrayBuffer[(Wishbone) => Unit]()
-
+  val requestAckCallbacks = ArrayBuffer[(Wishbone) => Unit]()
+  val responseCallbacks = ArrayBuffer[(Wishbone, WishboneTransaction, Boolean) => Unit]()
+  val requests = new Queue[(WishboneTransaction, Boolean)]()
   /** Add a callback, this will be executed on every bus acknoledge
     * @param callback a function or code block that takes a wishbone bus as parameter
     */
-  def addCallback(callback: (Wishbone) => Unit): Unit = callbacks += callback
+  def addCallback(callback: (Wishbone) => Unit): Unit = addReqAckCallback(callback)
```

Comment:
You sort of do need two callbacks for pipelined mode. You need one that captures the receive ACK host side -- essentially the address / we / wdata and then one that captures the receive response -- rdata. Since these can happen in different timesteps and in different order, it gets weird. I think the way I made it work now is reasonable -- capture the request ack, present it as the argument for the response callback -- but I didn't want to make current usages of the monitor break.

If I was designing it from scratch I'd keep the response callbacks and not have requestAck ones at all. Maybe the thing to do is revert out the changes to WishboneMonitor, make a WishbonePipelinedMonitor, and add that logic there. Maybe also deprecate the WishboneMonitor since it'd be more robust to test things in a pipeline aware way.

Reply:
Sorry, my initial thought here was just that if we shouldn't just add `addReqAckCallback` as an alias for `addCallback` and keep both.

TLDR; we can't really break users of the pipeline stuff since they pretty much can't exist ATM if I'm reading this correctly?

After looking into this more my 0.02:
- for standard the `addCallback` callback is called at the end of the transfer
- for pipelined `addCallback` doesn't make much sense as is, as you noted above its called to early - but: you couldn't have done anything useful for pipelined up until now since pipelined was broken in the rest of the lib. I'd therefore argue that you can't really break current users of the pipelined stuff.

What I don't like about the current proposal: We keep one function that's totally useful for standard, but not what you'd expect for pipelined. For the alias `addReqAckCallback`  I'd expect that an actual acknowledge is involved (before reading the source) - but that's only the case for standard, for pipelined its "not stalled" which is quite different. (request seems ambiguous here, is it the "address phase" or is it the whole read/write request - or is that a common term?)
The updated tests also suggest to use this function, but e.g. for `SpinalSimWishboneAdapterTester` it only work because it never does a read, same for the `SpinalSimWishboneArbiterTester`

We could just add an overload for addCallback so that the callback gets the needed data, and is always called at the end of a transfer? That would keep compatibility for existing code that works with standard mode, and do the expected thing for pipelined...
If we really fear breaking users of the pipelined stuff then we should think about making a non-broken version of the monitor that's not ambiguous to use. (not sure though it needs to be pipeline only)

@Dolu1990 any thoughts on this? 
I'd actually extend that thought to some of the "status" signals in the Wishbone bus itself

Reply:
I reworked this(including bringing back in clockdomain). I think this addresses your points and makes it make more sense?

If/when the implementation looks good I'll make sure none of the test code uses the older callback. 

If at some point we want an onError / onRetry those can also go into the trait but this PR has enough going on in it already